Installation issue: Boost linking on macOS fails, tries to use gcc for a clang build

### Steps to reproduce the issue

Minimal environment to reproduce the issue:
```yaml
spack:
  specs:
  - 'cmake@3.25.0:'
  - 'icu4c@76:'
  - 'boost@1.89.0:+json+locale+icu cxxstd=20'
  - 'hpx@1.11.0: cxxstd=20'
  view: true
  concretizer:
    unify: true
```

```console
$ spack spec -I boost
[+]  dcexj2r  boost@1.90.0~atomic~charconv~chrono~clanglibcpp~container~context~contract~conversion~date_time~debug~exception~fiber~filesystem~graph~graph_parallel+icu~iostreams+json+locale~log~math~mpi~mqtt5+multithreaded~nowide~numpy~openmethod~pic~program_options~python~random~regex~serialization+shared~signals2~singlethreaded~stacktrace~system~taggedlayout~test~thread~timer~type_erasure~url~versionedlayout~wave build_system=generic cxxstd=20 patches:=a440f96 visibility=hidden platform=darwin os=tahoe target=m3 %c,cxx=apple-clang@17.0.0
[e]  sdtgn2a      ^apple-clang@17.0.0 build_system=bundle platform=darwin os=tahoe target=aarch64
[+]  dwzpelv      ^compiler-wrapper@1.0 build_system=generic platform=darwin os=tahoe target=m3
[+]  as3xgwa      ^icu4c@76.1 build_system=autotools cxxstd=17 platform=darwin os=tahoe target=m3 %c,cxx=apple-clang@17.0.0
[e]  lra2mve          ^autoconf@2.73 build_system=autotools platform=darwin os=tahoe target=aarch64
[e]  d6hrae3          ^automake@1.18.1 build_system=autotools platform=darwin os=tahoe target=aarch64
[e]  c7xmboa          ^gmake@3.81~guile build_system=generic platform=darwin os=tahoe target=aarch64
[+]  petbq4m          ^gnuconfig@2025-07-10 build_system=generic platform=darwin os=tahoe target=m3
[e]  wwckkoq          ^libtool@2.5.4 build_system=autotools platform=darwin os=tahoe target=aarch64
[e]  nqvhfib          ^python@3.14.4+bz2+ctypes+dbm~debug~freethreading+libxml2+lzma~optimizations+pic+pyexpat~pythoncmd+readline+shared+sqlite3+ssl~static~tests~tkinter+uuid+zlib+zstd build_system=generic platform=darwin os=tahoe target=aarch64

$ spack find
==> In environment boost-hpx-debug (4 root specs)
 -  boost@1.89.0:+icu+json+locale cxxstd=20  [e] cmake@3.25.0:   -  hpx@1.11.0: cxxstd=20  [+] icu4c@76:

-- darwin-tahoe-aarch64 / no compilers --------------------------
apple-clang@17.0.0  autoconf@2.73  automake@1.18.1  cmake@4.3.2  git@2.54.0  gmake@3.81  libtool@2.5.4  ninja@1.13.2  pkgconf@2.5.1  python@3.14.4

-- darwin-tahoe-m3 / %c,cxx=apple-clang@17.0.0 ------------------
asio@1.32.0  gperftools@2.15  hwloc@2.13.0  icu4c@76.1  m4@1.4.21  ncurses@6.6  zlib-ng@2.3.3  zstd@1.5.7

-- darwin-tahoe-m3 / %c=apple-clang@17.0.0 ----------------------
bzip2@1.0.8  diffutils@3.12  libiconv@1.18  libsigsegv@2.15  libxml2@2.15.1  xz@5.8.3

-- darwin-tahoe-m3 / no compilers -------------------------------
compiler-wrapper@1.0  gnuconfig@2025-07-10
==> 26 installed packages
==> 2 concretized packages to be installed (show with `spack find -c`)
```


### Error message

<details><summary>Error message</summary>

<pre>
...
-- lines 17430 to 17450 --
  gcc.compile.c++ bin.v2/libs/math/build/gcc-21/release/arm_64/cxxstd-20-iso/link-static/pch-off/threading-multi/visibility-hidden/sph_legendrel.o
  gcc.compile.c++ bin.v2/libs/math/build/gcc-21/release/arm_64/cxxstd-20-iso/link-static/pch-off/threading-multi/visibility-hidden/sph_bessell.o
  gcc.compile.c++ bin.v2/libs/context/build/gcc-21/release/arm_64/cxxstd-20-iso/pch-off/threading-multi/visibility-hidden/posix/stack_traits.o
  gcc.compile.c++ bin.v2/libs/math/build/gcc-21/release/arm_64/cxxstd-20-iso/link-static/pch-off/threading-multi/visibility-hidden/sph_neumannl.o
  gcc.link.dll bin.v2/libs/context/build/gcc-21/release/arm_64/cxxstd-20-iso/pch-off/threading-multi/visibility-hidden/libboost_context.dylib
  ld: warning: -single_module is obsolete
> Undefined symbols for architecture arm64:
    "_jump_fcontext", referenced from:
        boost::context::detail::jump_fcontext(void*, void*) in fcontext.o
    "_make_fcontext", referenced from:
        boost::context::detail::make_fcontext(void*, unsigned long, void (*)(boost::context::detail::transfer_t)) in fcontext.o
    "_ontop_fcontext", referenced from:
        boost::context::detail::ontop_fcontext(void*, void*, boost::context::detail::transfer_t (*)(boost::context::detail::transfer_t)) in fcontext.o
  ld: symbol(s) not found for architecture arm64
> clang++: error: linker command failed with exit code 1 (use -v to see invocation)

      "/Users/jagot/prog/spack/opt/spack/darwin-m3/compiler-wrapper-1.0-dwzpelvzyhien3sc5hteapzbdisieh4j/libexec/spack/clang/clang++"    -o "bin.v2/libs/context/build/gcc-21/release/arm_64/cxxstd-20-iso/pch-off/threading-multi/visibility-hidden/libboost_context.dylib" -Wl,-install_name -Wl,"@rpath/libboost_context.dylib" -dynamiclib -single_module  "bin.v2/libs/context/build/gcc-21/release/arm_64/cxxstd-20-iso/pch-off/threading-multi/visibility-hidden/fcontext.o" "bin.v2/libs/context/build/gcc-21/release/arm_64/cxxstd-20-iso/pch-off/threading-multi/visibility-hidden/posix/stack_traits.o"       -fPIC -std=c++20 -fvisibility=hidden -fvisibility-inlines-hidden

  ...failed gcc.link.dll bin.v2/libs/context/build/gcc-21/release/arm_64/cxxstd-20-iso/pch-off/threading-multi/visibility-hidden/libboost_context.dylib...
  ...skipped <p/Users/jagot/prog/spack/opt/spack/darwin-m3/boost-1.90.0-sej3st62vv5fpf5exid2i3vacoyia7gs/lib>libboost_context.dylib for lack of <pbin.v2/libs/context/build/gcc-21/release/arm_64/cxxstd-20-iso/pch-off/threading-multi/visibility-hidden>libboost_context.dylib...
  gcc.compile.c++ bin.v2/libs/chrono/build/gcc-21/release/arm_64/cxxstd-20-iso/pch-off/threading-multi/visibility-hidden/process_cpu_clocks.o
==> Error: The following packages failed to install:
boost@1.90.0/sej3st62vv5fpf5exid2i3vacoyia7gs: /var/folders/3x/fs1m7b_d0458zrh1gt84s90c0000gn/T/jagot/spack-stage/spack-stage-boost-1.90.0-sej3st62vv5fpf5exid2i3vacoyia7gs-dol88ui9.log
</pre></details>


### Information on your system

```
spack debug report
* **Spack:** 1.2.0.dev0 (https://github.com/spack/spack/commit/d3ac7d4e2ad634fe2f9d6e5fa807b80c8388e7f0)
* **Builtin repo:** https://github.com/spack/spack-packages/commit/a8b4c96c86f389fd858484e901f310451633786f
* **Python:** 3.14.4
* **Platform:** darwin-tahoe-m3
```

### Additional information

@hainest 

[spack-build-out.txt](https://github.com/user-attachments/files/27032806/spack-build-out.txt)
[spack-build-env.txt](https://github.com/user-attachments/files/27032813/spack-build-env.txt)

## Workaround

For now, I can hack `package.py` to force the build to use the Clang compiler, the same way [it is done in Homebrew](https://github.com/Homebrew/homebrew-core/blob/92fc1b908dd50cdea63419b64073203af2a0a8ab/Formula/b/boost.rb#L46-L53):
```ruby
    # Force boost to compile with the desired compiler
    open("user-config.jam", "a") do |file|
      if OS.mac?
        file.write "using darwin : : #{ENV.cxx} ;\n"
      else
        file.write "using gcc : : #{ENV.cxx} ;\n"
      end
    end
```

### Hack

The hack is to modify `determine_toolset` to return `"clang"`:
```python
    def determine_toolset(self, spec):
        toolsets = {
            "%gcc": "gcc",
            "%intel": "intel",
            "%oneapi": "intel",
            "%clang": "clang",
            "%arm": "clang",
            "%xl": "xlcpp",
            "%xl_r": "xlcpp",
            "%nvhpc": "pgi",
            "%fj": "clang",
        }

        if spec.satisfies("@1.47:"):
            toolsets["%intel"] += "-linux"
            toolsets["%oneapi"] += "-linux"

        for cc, toolset in toolsets.items():
            print("{} {}: {}".format(cc, toolset, self.spec.satisfies(cc)))
            
        print(spec)

        return "clang"
```

For some reason, the spec does not satisfy `%clang`; `self.spec.satisfies` returns `False` for all the listed toolsets, even though `spack spec -l boost` clearly says `%c,cxx=apple-clang@17.0.0`:
```
%gcc gcc: False
%intel intel-linux: False
%oneapi intel-linux: False
%clang clang: False
%arm clang: False
%xl xlcpp: False
%xl_r xlcpp: False
%nvhpc pgi: False
%fj clang: False
```

### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [x] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

--- a/repos/spack_repo/builtin/packages/boost/package.py
+++ b/repos/spack_repo/builtin/packages/boost/package.py
@@ -544,6 +544,7 @@ class Boost(Package):
             "%intel": "intel",
             "%oneapi": "intel",
             "%clang": "clang",
+            "%apple-clang": "clang",
             "%arm": "clang",
             "%xl": "xlcpp",
             "%xl_r": "xlcpp",

